src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Specs/InterceptorInfo.cs

```diff
@@ -0,0 +1,197 @@
+                    ServiceCollectionExt = _interceptors_serviceCollectionExt?.OrderBy(i => i.FilePath).ToImmutableEquatableArray(),
+
+                    ConfigBinder_Bind_instance = _configBinder_InfoBuilder_Bind_instance?.ToIncrementalValue(),
+                    ConfigBinder_Bind_instance_BinderOptions = _configBinder_InfoBuilder_Bind_instance_BinderOptions?.ToIncrementalValue(),
+                    ConfigBinder_Bind_key_instance = _configBinder_InfoBuilder_Bind_key_instance?.ToIncrementalValue(),
+                };
+        }
+    }
+
+    internal sealed class TypeInterceptorInfoBuildler
+    {
+        private readonly Dictionary<ComplexTypeSpec, TypedInterceptorInvocationInfo.Builder> _invocationInfoBuilderCache = new();
+
+        public void RegisterInterceptor(MethodsToGen overload, ComplexTypeSpec type, IInvocationOperation invocation)
+        {
+            if (!_invocationInfoBuilderCache.TryGetValue(type, out TypedInterceptorInvocationInfo.Builder? invocationInfoBuilder))
+            {
+                _invocationInfoBuilderCache[type] = invocationInfoBuilder = new TypedInterceptorInvocationInfo.Builder(overload, type);
+            }
+
+            invocationInfoBuilder.RegisterInvocation(invocation);
+        }
+
+        public ImmutableEquatableArray<TypedInterceptorInvocationInfo>? ToIncrementalValue() =>
+            _invocationInfoBuilderCache.Values
+            .Select(b => b.ToIncrementalValue())
+            .OrderBy(i => i.TargetType.AssemblyQualifiedName)
+            .ToImmutableEquatableArray();
+    }
+
+    internal sealed record TypedInterceptorInvocationInfo
+    {
+        public required ComplexTypeSpec TargetType { get; init; }
+        public required ImmutableEquatableArray<InvocationLocationInfo> Locations { get; init; }
+
+        public void Deconstruct(out ComplexTypeSpec targetType, out ImmutableEquatableArray<InvocationLocationInfo> locations) =>
+            (targetType, locations) = (TargetType, Locations);
+
+        public sealed class Builder(MethodsToGen Overload, ComplexTypeSpec TargetType)
+        {
+            private readonly List<InvocationLocationInfo> _infoList = new();
+
+            public void RegisterInvocation(IInvocationOperation invocation) =>
+                _infoList.Add(new InvocationLocationInfo(Overload, invocation));
+
+            public TypedInterceptorInvocationInfo ToIncrementalValue() => new()
+            {
+                TargetType = TargetType,
+                Locations = _infoList.OrderBy(i => i.FilePath).ToImmutableEquatableArray()
+            };
+        }
+    }
+
+    internal sealed record InvocationLocationInfo
+    {
+        public InvocationLocationInfo(MethodsToGen interceptor, IInvocationOperation invocation)
+        {
+            MemberAccessExpressionSyntax memberAccessExprSyntax = ((MemberAccessExpressionSyntax)((InvocationExpressionSyntax)invocation.Syntax).Expression);
```

Comment:
Not entirely sure how safe these particular casts are. @333fred for evaluation.

Reply:
We do a similar cast in RDG as well, but this is after the following check has been invoked in the `predicate` of our SyntaxProvider. I assume that's OK?

https://github.com/dotnet/aspnetcore/blob/ee6d623741948c6ca907f60bad06736229b7d5e3/src/Http/Http.Extensions/gen/StaticRouteHandlerModel/InvocationOperationExtensions.cs#L73

Reply:
This isn't safe. Please add a test for invoking methods within the same type, ie `Method(parameter)` from within the same static class.

Reply:
Agree it should be safe; the invocation has already been validated upstream. Can make this clearer with `Debug.Assert` and/or a comment.

Reply:
Fred points out that the cast is not valid if the method is invoked within the extension class, without syntax to reference the class.

Not new in this PR; will address in follow up. Also it's unlikely that the source generator will run on the same methods that it is intercepting.